insecure-time/src/lib.rs

```diff
@@ -0,0 +1,853 @@
+}
+
+impl Ticks {
+    pub fn new(t: u64) -> Self {
+        Ticks(AtomicU64::new(t))
+    }
+
+    pub fn now() -> Self {
+        Ticks(Rdtscp::read().into())
+    }
+
+    pub fn abs_diff(&self, t1: &Ticks) -> Ticks {
+        let t0 = self.0.load(Ordering::Relaxed);
+        let t1 = t1.0.load(Ordering::Relaxed);
+
+        if t1 > t0 {
+            Ticks::new(t1 - t0)
+        } else {
+            Ticks::new(t0 - t1)
+        }
+    }
+
+    fn set(&self, t: Ticks) {
+        let t = t.0.load(Ordering::Relaxed);
+        self.0.store(t, Ordering::Relaxed);
+    }
+
+    pub fn from_duration(duration: Duration, freq: &Freq) -> Self {
+        let freq = freq.as_u64();
+        let ticks_secs = duration.as_secs() * freq;
+        let ticks_nsecs = duration.subsec_nanos() as u64 * freq / NANOS_PER_SEC as u64;
+        Ticks::new(ticks_secs + ticks_nsecs)
+    }
+
+    pub fn as_duration_ex(&self, freq: &Freq) -> Duration {
+        let freq = freq.as_u64();
+        let ticks = self.0.load(Ordering::Relaxed);
+
+        let time_secs = ticks / freq;
+        let time_nsecs = (ticks % freq * NANOS_PER_SEC as u64) / freq;
+        let time_nsecs: u32 = time_nsecs.try_into().expect("must be smaller than 1sec");
+
+        Duration::new(time_secs, time_nsecs)
+    }
+
+    pub fn elapsed(&self, freq: &Freq) -> Duration {
+        let tsc_now = Ticks::now();
+        self.abs_diff(&tsc_now).as_duration_ex(&freq)
+    }
+}
+
+pub struct Rdtscp;
+
+impl Rdtscp {
+    #[inline(never)]
+    pub fn read() -> u64 {
+        let mut aux: u32 = 0;
+        unsafe { __rdtscp(&mut aux) }
+    }
+
+    #[cfg(not(target_env = "sgx"))]
+    pub fn is_supported() -> bool {
+        let mut eax = 0;
+        let mut ebx = 0;
+        let mut ecx = 0;
+        let mut edx = 0;
+
+        cpuid(0x12, &mut eax, &mut ebx, &mut ecx, &mut edx);
+        eax & (0x1 << 1) != 0x0
+
+    }
+}
+
+pub struct Freq(AtomicU64);
+
+impl PartialEq<Freq> for Freq {
+    fn eq(&self, other: &Freq) -> bool {
+        let f = self.0.load(Ordering::Relaxed);
+        let o = other.0.load(Ordering::Relaxed);
+        f.eq(&o)
+    }
+}
+
+impl PartialOrd<Freq> for Freq {
+    fn partial_cmp(&self, other: &Freq) -> Option<cmp::Ordering> {
+        let f = self.0.load(Ordering::Relaxed);
+        let o = other.0.load(Ordering::Relaxed);
+        f.partial_cmp(&o)
+    }
+}
+
+impl Debug for Freq {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), fmt::Error> {
+        let freq = self.as_u64();
+        f.debug_struct("Freq")
+         .field("freq", &freq)
+         .finish()
+    }
+}
+
+impl Freq {
+    fn is_zero(&self) -> bool {
+        self.0.load(Ordering::Relaxed) == 0
+    }
+
+    fn set(&self, f: &Freq) {
+        let f = f.0.load(Ordering::Relaxed);
+        self.0.store(f, Ordering::Relaxed);
+    }
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum ClockSource {
+    TSC,
+}
+
+impl Freq {
+    pub fn new(crystal_hz: u32, numerator: u32, denominator: u32) -> Self {
```

Comment:
since we're dividing and computing remainders by `Freq` in a lot of places, maybe we should ensure this is never 0

Reply:
Nevermind, I just saw the `assert!(freq != 0);` in line 228 below.